### PR TITLE
Fixes #24427 - Modules available to a Host

### PR DESCRIPTION
**Pull request by parthaa**
Added API constructs to show modules that are available to a host based on the
enabled repositories in the host.

### CODE REVIEW

**Comment by parthaa** [comment]
```
$curl -X GET -s -k -u admin:changeme 'https://localhost/katello/api/v2/module_streams?host_ids[]=1'  --globoff
{"total":7,"subtotal":7,"page":1,"per_page":20,"error":null,"search":null,"sort":{"by":"name","order":"asc"},"results":[{"id":7,"name":"duck","uuid":"c546685e-199e-4d6a-89ea-9dad337a16e8","version":"20180704244205","context":"deadbeef","stream":"0","arch":"noarch","description":"A module for the duck 0.6 package","summary":"Duck 0.6 module","repositories":[{"id":1,"name":"modular"},{"id":2,"name":"modular"}]},{"id":6,"name":"duck","uuid":"64d206bc-5034-4215-8171-ded97143323a","version":"201809302113907","context":"deadbeef","stream":"0","arch":"noarch","description":"A module for the duck 0.8 package","summary":"Duck 0.8 module","repositories":[{"id":1,"name":"modular"},{"id":2,"name":"modular"}]},{"id":1,"name":"duck","uuid":"191a36b4-8d6d-460b-a16c-c3bd0851724f","version":"20180730233102","context":"deadbeef","stream":"0","arch":"noarch","description":"A module for the duck 0.7 package","summary":"Duck 0.7 module","repositories":[{"id":1,"name":"modular"},{"id":2,"name":"modular"}]},{"id":4,"name":"kangaroo","uuid":"47878418-b635-4425-88c5-159199d3fb7b","version":"20180704111719","context":"deadbeef","stream":"0","arch":"noarch","description":"A module for the kangaroo 0.2 package","summary":"Kangaroo 0.2 module","repositories":[{"id":1,"name":"modular"},{"id":2,"name":"modular"}]},{"id":3,"name":"kangaroo","uuid":"45bc4a06-78e7-4ca5-ad60-31fe45f67ef6","version":"20180730223407","context":"deadbeef","stream":"0","arch":"noarch","description":"A module for the kangaroo 0.3 package","summary":"Kangaroo 0.3 module","repositories":[{"id":1,"name":"modular"},{"id":2,"name":"modular"}]},{"id":5,"name":"walrus","uuid":"63cadf68-3189-44d6-b6a5-af3d8d52b02b","version":"20180704144203","context":"deadbeef","stream":"5.21","arch":"x86_64","description":"A module for the walrus 5.21 package","summary":"Walrus 5.21 module","repositories":[{"id":1,"name":"modular"},{"id":2,"name":"modular"}]},{"id":2,"name":"walrus","uuid":"411bff5d-a63d-456b-8d0e-540e94dc1c13","version":"20180707144203","context":"c0ffee42","stream":"0.71","arch":"x86_64","description":"A module for the walrus 0.71 package","summary":"Walrus 0.71 module","repositories":[{"id":1,"name":"modular"},{"id":2,"name":"modular"}]}]}
```

**Comment by parthaa** [comment]
Note: at the present time we do not have the plugin that will report the enabled repositories back to the server for either fedora 28/rhel 8 so to really test this I had to artificially create a an entry in ::Katello::ContentFacetRepository by doing ```::Katello::ContentFacetRepository.create!(:content_facet_id =><..>, :repository_id => <..>)```

File: app/controllers/katello/api/v2/module_streams_controller.rb [hunk 1]
```diff
@@ -3,8 +3,43 @@ class Api::V2::ModuleStreamsController < Api::V2::ApiController
     apipie_concern_subst(:a_resource => N_("a module stream"), :resource => "module_streams")
     include Katello::Concerns::Api::V2::RepositoryContentController
 
+    before_action :check_hosts, :only => :index
+
+    api :GET, "/module_streams", N_("List packages")
+    api :GET, "/content_views/:content_view_id/filters/:filter_id/:resource_id", N_("List :resource_id")
+    api :GET, "/content_view_filters/:content_view_filter_id/:resource_id", N_("List :resource_id")
+    api :GET, "/repositories/:repository_id/:resource_id", N_("List :resource_id")
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/module_streams_controller.rb]: I suppose since you are overriding the parent class' index method, you need to redeclare the api docs?

File: app/controllers/katello/api/v2/module_streams_controller.rb [hunk 2]
```diff
@@ -3,8 +3,43 @@ class Api::V2::ModuleStreamsController < Api::V2::ApiController
...
+    param :content_view_version_id, :number, :desc => N_("content view version identifier")
+    param :content_view_filter_id, :number, :desc => N_("content view filter identifier")
+    param :repository_id, :number, :desc => N_("repository identifier")
+    param :environment_id, :number, :desc => N_("environment identifier")
+    param :ids, Array, :desc => N_("ids to filter content by")
+    param :host_ids, Array, :desc => N_("List of host id to list available module streams for")
+    param_group :search, ::Katello::Api::V2::ApiController
+    def index
+      super
+    end
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/module_streams_controller.rb]: Do you need this method? I think you can just define custom_index_relation

Comment (johnpmitsch) [reply app/controllers/katello/api/v2/module_streams_controller.rb]: I guess you are doing this to add `host_ids` param?

File: app/controllers/katello/api/v2/module_streams_controller.rb [hunk 3]
```diff
@@ -3,8 +3,43 @@ class Api::V2::ModuleStreamsController < Api::V2::ApiController
...
+
+    def custom_index_relation(collection)
+      if params[:host_ids]
+        collection = collection.available_for_hosts(params[:host_ids])
+      end
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/module_streams_controller.rb]: probably could make this a one liner

File: app/models/katello/module_stream.rb [hunk 4]
```diff
@@ -50,5 +50,15 @@ def create_profiles(profiles)
...
+
+    def self.available_for_hosts(hosts)
+      joins("INNER JOIN #{Katello::RepositoryModuleStream.table_name} on \
+        #{Katello::RepositoryModuleStream.table_name}.module_stream_id = #{Katello::ModuleStream.table_name}.id").
+      joins("INNER JOIN #{Katello::ContentFacetRepository.table_name} on \
+        #{Katello::RepositoryModuleStream.table_name}.repository_id = #{Katello::ContentFacetRepository.table_name}.id").
+      joins("INNER JOIN #{Katello::Host::ContentFacet.table_name} on \
+        #{Katello::Host::ContentFacet.table_name}.id = #{Katello::ContentFacetRepository.table_name}.content_facet_id").
+      where("#{Katello::Host::ContentFacet.table_name}.host_id" => hosts).distinct
+    end
```
Comment (johnpmitsch) [inline app/models/katello/module_stream.rb]: You have a couple options in AR for this (assuming I'm understanding the query correctly)
```ruby
Katello::ModuleStream.joins(repositories: :content_facets).where('katello_content_facets.host_id' => [4])
# OR
Katello::ModuleStream.joins(repositories: :content_facets).merge( Katello::Host::ContentFacet.where(host_id: [4]) )
```

File: app/controllers/katello/api/v2/module_streams_controller.rb (hunk 2 above)
Comment (parthaa) [reply app/controllers/katello/api/v2/module_streams_controller.rb]: yup I needed to add the host_ids param. Thats the only reason

File: app/controllers/katello/api/v2/module_streams_controller.rb (hunk 1 above)
Comment (parthaa) [reply app/controllers/katello/api/v2/module_streams_controller.rb]: I need it to add host_ids

File: app/models/katello/module_stream.rb [hunk 5]
```diff
@@ -50,5 +50,10 @@ def create_profiles(profiles)
         end
       end
     end
+
+    def self.available_for_hosts(hosts)
+      joins(repositories: :content_facets).merge(Katello::Host::ContentFacet.where(host_id: hosts)).distinct
+
```
Comment (johnpmitsch) [inline app/models/katello/module_stream.rb]: extra blank line :upside_down_face:

File: app/controllers/katello/api/v2/module_streams_controller.rb [hunk 6]
```diff
@@ -3,8 +3,44 @@ class Api::V2::ModuleStreamsController < Api::V2::ApiController
     apipie_concern_subst(:a_resource => N_("a module stream"), :resource => "module_streams")
     include Katello::Concerns::Api::V2::RepositoryContentController
 
+    before_action :check_hosts, :only => :index
+
+    api :GET, "/module_streams", N_("List packages")
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/module_streams_controller.rb]: You have "List packages" here instead of module streams

File: app/controllers/katello/api/v2/module_streams_controller.rb [hunk 7]
```diff
@@ -3,8 +3,44 @@ class Api::V2::ModuleStreamsController < Api::V2::ApiController
...
+    api :GET, "/content_views/:content_view_id/filters/:filter_id/:resource_id", N_("List :resource_id")
+    api :GET, "/content_view_filters/:content_view_filter_id/:resource_id", N_("List :resource_id")
+    api :GET, "/repositories/:repository_id/:resource_id", N_("List :resource_id")
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/module_streams_controller.rb]: What does this endpoint do? `List :resource_id` isn't that clear

**Comment by johnpmitsch** [comment]
@iNecas the controller in this PR (module_streams_controller.rb) extends [the index method in repository_content_controller](https://github.com/Katello/katello/blob/master/app/controllers/katello/concerns/api/v2/repository_content_controller.rb#L30) in order to add one param to said controller. To add this param, all the apipie methods have to be redeclared and a dummy index method that just called super is added to the child class. Is there a better way to do this?

tl;dr trying to add a single param to parent class's apipie definitions

**Comment by iNecas** [comment]
Have a look at https://github.com/Apipie/apipie-rails#example-11

**Comment by omkarkhatavkar** [comment]
**Steps Executed :**

1. Create module streams using reos ( , )
2. Created Life-Cycle Environment.
3. Created Content View and added repos and promoted to environment
4. Created activation key with autoattched enabled and add the subscribed all products.
5. Link Content Host (fedora28) with activation key 

```
[root@fedora28 vagrant]# yum repolist --enabled
Last metadata expiration check: 0:00:23 ago on Tue 18 Sep 2018 07:29:48 AM UTC.
repo id                                                                  repo name                                      status
Default_Organization_Product_1_Modularity_animal_modules                 animal modules                                     34
Default_Organization_Product_1_Modularity_duplicate_animal_repo          duplicate animal repo                              34
Default_Organization_Product_2_Modularity_dev_modules                    dev_modules                                         9
```

6. Run dnf command -> to enable the plugin 
`dnf install screen -y `
 
```
[vagrant@centos7-devel foreman]$ rails c
Running via Spring preloader in process 12994
2018-09-18T07:31:47 [D|app|] Current user set to foreman_console_admin (admin)
2018-09-18T07:31:47 [W|app|] Console started with 'foreman_console_admin' user, call User.current= to change it
Loading development environment (Rails 5.2.1)
irb(main):001:0> ::Katello::ContentFacetRepository.all
+----+------------------+---------------+
| id | content_facet_id | repository_id |
+----+------------------+---------------+
| 34 | 3                | 17            |
| 35 | 3                | 18            |
| 36 | 3                | 19            |
+----+------------------+---------------+
3 rows in set
irb(main):002:0> exit()
[vagrant@centos7-devel foreman]$ curl -X GET -s -k -u admin:changeme 'https://localhost/katello/api/v2/module_streams?host_ids[]=3'  --globoff 
{"total":0,"subtotal":0,"page":1,"per_page":20,"error":null,"search":null,"sort":{"by":"name","order":"asc"},"results":[]}
[vagrant@centos7-devel foreman]$ 
```
**Result**: 
**Fails to give response.** 

**Passed Tests** 
```
[vagrant@centos7-devel foreman]$ curl -X GET -s -k -u admin:changeme 'https://localhost/katello/api/v2/module_streams?host_ids[]=invalid'  --globoff 
{"displayMessage":"One or more hosts not found","errors":["One or more hosts not found"]}[vagrant@centos7-devel foreman]$ 
```

**Comment by johnpmitsch** [comment]
@iNecas thanks! apipie.docs++ 

@parthaa looks like you can use the `update_api` method

File: app/controllers/katello/api/v2/module_streams_controller.rb (hunk 7 above)
Comment (parthaa) [reply app/controllers/katello/api/v2/module_streams_controller.rb]: just copied the super method api's.  check packages_controller which does the same thing

**Comment by parthaa** [comment]
> **Result**:
> **Fails to give response.**

 @omkarkhatavkar - It could be the case that you used the content_facet_id  = 3 for host_ids 
What we are looking for is content_facet.host_id , find the host id for your host and use that

File: app/models/katello/module_stream.rb [hunk 8]
```diff
@@ -50,5 +50,10 @@ def create_profiles(profiles)
         end
       end
     end
+
+    def self.available_for_hosts(hosts)
```
[the remark this answers is not in the document]
Comment (parthaa) [reply app/models/katello/module_stream.rb]: addressed

**Comment by parthaa** [comment]
> ;dr trying to add a single param to parent class's apipie definitions

addressed

File: app/controllers/katello/api/v2/module_streams_controller.rb [hunk 9]
```diff
@@ -3,8 +3,35 @@ class Api::V2::ModuleStreamsController < Api::V2::ApiController
...
 
+    before_action :check_hosts, :only => :index
+
+    update_api(:index) do
+      param :host_ids, Array, :desc => N_("List of host id to list available module streams for")
+    end
+
+    def index
+      super
+    end
```
Comment (johnpmitsch) [inline app/controllers/katello/api/v2/module_streams_controller.rb]: Do you still  need this index method? I would think its not necessary anymore with the addition of `update_api` method above

**Comment by omkarkhatavkar** [comment]
@parthaa : Still not working :(
**1. hammer host list shows id is 5** 

`hammer> host list
---|-------------------------------|------------------|------------|-----------------|-------------------|---------------|----------------------
ID | NAME                          | OPERATING SYSTEM | HOST GROUP | IP              | MAC               | CONTENT VIEW  | LIFECYCLE ENVIRONMENT
---|-------------------------------|------------------|------------|-----------------|-------------------|---------------|----------------------
5  | fedora28.okhatavk.example.com | Fedora 28        |            | 192.168.121.114 | 52:54:00:62:2d:60 | Content_View2 | Library              
---|-------------------------------|------------------|------------|-----------------|-------------------|---------------|----------------------
hammer> 
`

**2. Getting data from rails c** 
```

[vagrant@centos7-devel foreman]$ rails c 
Running via Spring preloader in process 17193
2018-09-19T06:46:39 [D|app|] Current user set to foreman_console_admin (admin)
2018-09-19T06:46:39 [W|app|] Console started with 'foreman_console_admin' user, call User.current= to change it
Loading development environment (Rails 5.2.1)
irb(main):001:0> ::Katello::ContentFacetRepository.all
+----+------------------+---------------+
| id | content_facet_id | repository_id |
+----+------------------+---------------+
| 48 | 5                | 21            |
| 49 | 5                | 23            |
+----+------------------+---------------+
2 rows in set
irb(main):002:0> exit()
[vagrant@centos7-devel foreman]$ curl -X GET -s -k -u admin:changeme 'https://localhost/katello/api/v2/module_streams?host_ids[]=5'  --globoff 
{"total":0,"subtotal":0,"page":1,"per_page":20,"error":null,"search":null,"sort":{"by":"name","order":"asc"},"results":[]}
[vagrant@centos7-devel foreman]$ 
```

**3. See the module list and repolist on content host** 

```
[root@fedora28 ~]# yum repolist
Last metadata expiration check: 0:00:11 ago on Wed 19 Sep 2018 06:53:08 AM UTC.
repo id                                                                  repo name                                      status
Default_Organization_Product_1_Modularity_animal_modules                 animal modules                                     34
Default_Organization_Product_1_Modularity_duplicate_animal_repo          duplicate animal repo                              34
*fedora                                                                  Fedora 28 - x86_64                             57,321
*fedora-modular                                                          Fedora Modular 28 - x86_64                          9
rhel-client                                                              rhel-client                                         5
*updates                                                                 Fedora 28 - x86_64 - Updates                   18,871
*updates-modular                                                         Fedora Modular 28 - x86_64 - Updates                6
[root@fedora28 ~]# 
```
(i tried multiple time to register and unregister content hosts with dnf install screen (just to start the dnf plugin))
Please let me know your thoughts on this.

**Comment by parthaa** [comment]
@omkarkhatavkar should be fixed now. Was an issue with publish. So you can Either
1. Run ```Katello::Repository.linked_repositories.each(&:index_content)``` OR
1. Republish the CV with this pr and try again
1. Note you can run ```dnf katello-upload-enabled-repos``` to force the upload instead of ```dnf install <..> ```

**Comment by parthaa** [comment]
[test katello]

**Comment by omkarkhatavkar** [comment]
@parthaa, I was not able to publish the repo, getting below error while publishing content view. 
```
PG::NotNullViolation: ERROR:  null value in column "created_at" violates not-null constraint
DETAIL:  Failing row contains (179, 38, 1, null, null).
: insert into katello_repository_module_streams (repository_id, module_stream_id)
                        select 38 as repository_id, module_stream_id from katello_repository_module_streams
                        where repository_id = 37 and module_stream_id not in (select module_stream_id
                        from katello_repository_module_streams where repository_id = 38)
```
running `Run Katello::Repository.linked_repositories.each(&:index_content) `command on rails console gives the same error. 
Do you want me to do reset db and perform all operations again, if it is DB issue ?

**Comment by parthaa** [comment]
> running `Run Katello::Repository.linked_repositories.each(&:index_content)`command on rails console gives the same error.

Weird I am now hitting this error as well. Might be something off with my test environment. I added a migration to fix this. Hopefully that will do the trick

**Comment by johnpmitsch** [comment]
Works!  Had to file an unrelated issue I was able to workaround with hammer https://projects.theforeman.org/issues/24999

**Comment by parthaa** [comment]
[test katello]

**Comment by parthaa** [comment]
[test katello]

**Comment by parthaa** [comment]
[test katello]
